clang/lib/Sema/HLSLExternalSemaSource.cpp

```diff
@@ -323,30 +324,119 @@ struct TemplateParameterListBuilder {
         S.Context, Builder.Record->getDeclContext(), SourceLocation(),
         SourceLocation(), /* TemplateDepth */ 0, Position,
         &S.Context.Idents.get(Name, tok::TokenKind::identifier),
-        /* Typename */ false,
-        /* ParameterPack */ false);
+        /* Typename */ true,
+        /* ParameterPack */ false,
+        /* HasTypeConstraint*/ false);
     if (!DefaultValue.isNull())
       Decl->setDefaultArgument(
           S.Context, S.getTrivialTemplateArgumentLoc(DefaultValue, QualType(),
                                                      SourceLocation()));
-
+    Decl->setReferenced();
     Params.emplace_back(Decl);
     return *this;
   }
 
-  BuiltinTypeDeclBuilder &finalizeTemplateArgs() {
+  /*
+  The desired result AST after this function constructs the concept
+  specialization Expression is as follows:
+
+  |-ConceptSpecializationExpr 0xdd5d408 <col:42, col:75> 'bool' Concept
```

Comment:
Did some autoformatter do something bad with the formatting here?

Reply:
Yes, this was meant to show the AST structure.